src/server/pps/server/poller.go

```diff
@@ -119,27 +147,133 @@ func (a *apiServer) pollPipelines(pachClient *client.APIClient) {
+		// prevents pachyderm from creating pipelines when there's an issue
+		// talking to k8s.
+		kubePipelineWatch, err := m.a.env.GetKubeClient().CoreV1().Pods(m.a.namespace).Watch(
+			metav1.ListOptions{
+				LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(
+					map[string]string{
+						"component": "worker",
+					})),
+				Watch: true,
+			})
+		if err != nil {
+			return errors.Wrap(err, "failed to watch kubernetes pods")
+		}
+		defer kubePipelineWatch.Stop()
+		for event := range kubePipelineWatch.ResultChan() {
+			// if we get an error we restart the watch
+			if event.Type == kube_watch.Error {
+				return errors.Wrap(kube_err.FromObject(event.Object), "error while watching kubernetes pods")
+			} else if event.Type == "" {
+				// k8s watches seem to sometimes get stuck in a loop returning events
+				// with Type = "". We treat these as errors as otherwise we get an
+				// endless stream of them and can't do anything.
+				return errors.New("error while watching kubernetes pods: empty event type")
+			}
+			pod, ok := event.Object.(*v1.Pod)
+			if !ok {
+				continue // irrelevant event
+			}
+			if pod.Status.Phase == v1.PodFailed {
+				log.Errorf("pod failed because: %s", pod.Status.Message)
+			}
+			pipelineName := pod.ObjectMeta.Annotations["pipelineName"]
+			for _, status := range pod.Status.ContainerStatuses {
+				if status.State.Waiting != nil && failures[status.State.Waiting.Reason] {
+					if err := m.a.setPipelineCrashing(ctx, pipelineName, status.State.Waiting.Message); err != nil {
+						return errors.Wrap(err, "error moving pipeline to CRASHING")
+					}
+				}
+			}
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == v1.PodScheduled &&
+					condition.Status != v1.ConditionTrue && failures[condition.Reason] {
+					if err := m.a.setPipelineCrashing(ctx, pipelineName, condition.Message); err != nil {
+						return errors.Wrap(err, "error moving pipeline to CRASHING")
+					}
+				}
+			}
+		}
+		return backoff.ErrContinue // keep polling until cancelled (RetryUntilCancel)
+	}, &backoff.ZeroBackOff{}, backoff.NotifyContinue("pollPipelinePods"),
+	); err != nil && ctx.Err() == nil {
+		panic("pollPipelinePods is exiting prematurely which should not happen; restarting pod...")
+	}
+}
+
+// pollPipelinesEtcd watches the 'pipelines' collection in etcd and sends
+// writeEv and deleteEv events to the PPS master when it sees them.
+//
+// pollPipelinesEtcd is unlike the other poll and monitor goroutines in that it
+// sees the result of other poll/monitor goroutines' writes. For example, when
+// pollPipelinePods (above) observes that a pipeline is crashing and updates its
+// state in etcd, the flow for starting monitorPipelineCrashing is:
+//
+//  k8s watch ─> pollPipelinePods  ╭──> pollPipelinesEtcd  ╭──> m.run()
+//                      │          │            │          │      │
+//                      ↓          │            ↓          │      ↓
```

Comment:
Between this and the invariant in that createBranch comment, I wonder if you have the most unique unicode codepoints across commits.

Reply:
:bow: Thank you for your kind words. As soon as I heard Go supported unicode source files, I knew what my calling was in life.